src/github.com/getlantern/detour/detour.go

```diff
@@ -1,400 +1,162 @@
-4. Connection made but get fake response or HTTP redirect to a fixed URL.
-
-[1] Detour will not replay nonidempotent plain HTTP requests, but will add it to
-    proxied list to be detoured next time.
-[2] Detour can only handle exact 1 successful read followed by failed read,
-    which covers most cases in reality.
-*/
 package detour
 
 import (
-	"bytes"
 	"fmt"
 	"net"
 	"sync"
 	"sync/atomic"
 	"time"
 
+	"github.com/getlantern/eventual"
 	"github.com/getlantern/golog"
 )
 
-// If no any connection made after this period, stop dialing and fail
-var TimeoutToConnect = 30 * time.Second
-
-// To avoid unnecessarily proxy not-blocked url, detour will dial detour connection
-// after this small delay. Set to zero to dial in parallel to not introducing any delay.
-var DelayBeforeDetour = 0 * time.Millisecond
-
-// If DirectAddrCh is set, when a direct connection is closed without any error,
-// the connection's remote address (in host:port format) will be send to it
-var DirectAddrCh = make(chan string)
-
 var (
 	log = golog.LoggerFor("detour")
-)
-
-// Conn implements an net.Conn interface by utilizing underlie direct and
-// detour connections.
-type Conn struct {
-	// Keeps track of the total bytes read from this connection, atomic
-	// Due to https://golang.org/pkg/sync/atomic/#pkg-note-BUG it requires
-	// manual alignment. For this, it is best to keep it as the first field
-	readBytes uint64
-
-	// The underlie connections, uses buffered channel as ring queue to avoid
-	// locking. We have at most 2 connetions so a length of 2 is enough.
-	conns chan conn
-
-	// The chan to notify dialer to dial detour immediately
-	chDialDetourNow chan bool
-	// The channel to notify read/write that a detour connection is available
-	chDetourConn chan conn
-
-	// The chan to receive result of any read operation
-	chRead chan ioResult
-	// The chan to receive result of any write operation
-	chWrite chan ioResult
-
-	addr string
-
-	muWriteBuffer sync.RWMutex
-	// Keeps written bytes through direct connection to replay it if required.
-	writeBuffer *bytes.Buffer
-	// Is it a plain HTTP request or not, atomic
-	nonidempotentHTTPRequest uint32
-}
 
-// The data structure to pass result of io operation back from underlie connection
-type ioResult struct {
-	// Number of bytes read/wrote
-	n int
-	// IO error, if any
-	err error
-	// The underlie connection itself
-	conn conn
-}
-
-type connType int
-
-const (
-	connTypeDirect connType = iota
-	connTypeDetour connType = iota
+	// DirectDialTimeout controls the dial timeout for direct connections
+	// We make this relatively small because detour won't fail over to a proxied
+	// connection until this timeout is hit. The detoured connection is eagerly
+	// fetched, so once we fail over the request will complete very quickly, so
+	// 5 seconds is a good approximation of our maximum latency for initial
+	// detouring.
+	DirectDialTimeout = 5 * time.Second
```

Comment:
We probably don't want to introduce timeout, but use whichever connection responds first. If detour connection responds in 1s, waiting for another 4s is unnecessary, but in extremely slow network (GPRS?), it's possible that a direct accessible site takes 10s to respond (and detour connection may take longer). That's the main reason we want to improve detour.

The idea of buffered connection is very inspirational. Having always available connections should eliminate most work to coordinate `Dial`, `Read` and `Write`.